### PR TITLE
player: always write redirect entries for resuming playback

### PR DESCRIPTION
**Pull request by guidocella**
35f43dfacbe added a system to write resume files for redirects, i.e. directories and playlists that mpv expands.

It creates a resume file for each redirect, and for the first redirect only, it writes a resume file for each segment of its path, without even converting it to an absolute path if it's relative. This is incomplete and overcomplicated and it's strange that wm4 went for it:

`mpv 'Iron Maiden/1982 The Number of the Beast/8 Hallowed Be Thy Name.mp3'`
This doesn't save any redirect entry.

`mpv --directory-mode=recursive 'Iron Maiden'`, then quit-watch-later on Hallowed Be Thy Name
This saves a redirect entry for "Iron Maiden", but not for "1982 The Number of the Beast". It doesn't save redirect entries for the directories above "Iron Maiden" either because "Iron Maiden" isn't converted to an absolute path.

In both of these cases `mpv --directory-mode=lazy 'Iron Maiden'` won't resume from "Hallowed Be Thy Name" because "1982 The Number of the Beast" isn't the first subdirectory and there is no redirect entry for it.

503dada42f made mpv recursively expand subdirectories precisely to fix this, and f266eadf1e added back an option not to expand them. But if we fix how redirect entries are stored, we can make the superior `--directory-mode=lazy` (because it's faster and doesn't result in massive playlists) the default, and also ensure that mpv will resume playback even when you quit-watch-later a file without redirects and then play the directories above it.

To fix this just ignore redirects and always create redirect entries for all segments of the absolute path of the file, so that both

`mpv 'Iron Maiden/1982 The Number of the Beast/8 Hallowed Be Thy Name.mp3'`
`mpv --directory-mode=lazy 'Iron Maiden'`

will create redirect entries for

/$USER
/$USER/music
/$USER/music/Iron Maiden
/$USER/music/Iron Maiden/1982 The Number of the Beast

making `mpv --directory-mode=lazy "Iron Maiden"` resume from "Hallowed Be Thy Name".

The only case where creating resume files specifically for redirected playlists is useful is when you have a playlist pointing to files in different parent directories, e.g. /foo/bar.m3u => /baz/qux.mkv, and then do `mpv quux.mkv /foo/bar.m3u`. This should be very uncommon and therefore not worth adding code for. It doesn't look like wm4 was thinking of this use case either from 35f43dfacbe's git message and code comments.

The code to track redirects could now be removed altogether if it wasn't for `player/loadfile.c` checking `mpctx->playing->num_redirects` to determine whether to log the filename when playing a file in a directory with a single file.

### CODE REVIEW

**Comment by Dudemanguy** [comment]
I do like the new behavior, but this seems to leave extraneous (?) redirect entries for some reason. Like `mpv foo` -> shift+q -> `mpv foo` and there's still stuff in the watch_later directory.

File: player/configfiles.c
Comment (Dudemanguy) [inline player/configfiles.c:388]: I got a segfault from this line. Seems like `fname` can was freed a few lines up?

Comment (guidocella) [reply player/configfiles.c]: I was using the wrong variable. I wonder why I never got a segfault.

Comment (Dudemanguy) [reply player/configfiles.c]: Ah yeah there we go.

**Comment by Dudemanguy** [comment]
Just needs a squash but this seems good to me now.

**Comment by guidocella** [comment]
The one use case that came to my mind where writing resume files for redirects like master currently does may be useful is when playing multiple archives of images, with this patch a watch later file is created for the archive://... image but not a redirect entry for the archive file itself. I'm not sure if it's worth keeping both systems just for that and if anyone actually uses this.

**Comment by natural-harmonia-gropius** [comment]
**Additional request:**

Currently `a/b/c` and `a\b\c` are treated as different paths, replace `\` with `/` will be good.

https://github.com/natural-harmonia-gropius/recent-menu/issues/15

**Comment by guidocella** [comment]
I plan to normalize paths in a later PR. There are many more problematic cases like foo.mkv, ./foo.mkv .//foo.mkv, ../"$(basename $PWD)"/foo.mkv all counting as different paths.

**Comment by Dudemanguy** [comment]
>The one use case that came to my mind where writing resume files for redirects like master currently does may be useful is when playing multiple archives of images

I tested this for a bit and didn't notice any changes in behavior. i.e. it still resumed from the appropriate spot in the latter archives. I was only using two zips or archives though. Not sure if that's the same scenario you were thinking of.

**Comment by guidocella** [comment]
> I tested this for a bit and didn't notice any changes in behavior. i.e. it still resumed from the appropriate spot in the latter archives. I was only using two zips as archives though. Not sure if that's the same scenario you were thinking of.

`mpv 1.zip 2.zip`, `quit-watch-later` on a file in `2.zip`, `mpv 1.zip 2.zip` does not resume from `2.zip`, though once you play `2.zip`, it will resume from the image you quit from.

**Comment by Dudemanguy** [comment]
>mpv 1.zip 2.zip does not resume from 2.zip, though once you play 2.zip

Oh I see, it doesn't work if you use `--directory-mode=lazy` with this PR but it does on master.

**Comment by Dudemanguy** [comment]
Nice so using playlist-path seems to have fixed the limitation you mentioned earlier. I think this is good to go unless you can think of something else.

File: player/configfiles.c
Comment (Dudemanguy) [inline player/configfiles.c:385]: I think we can remove this comment.

**Review by Dudemanguy — APPROVED** [review]
I'll play with a bit more before merging but LGTM.

**Pull request merged**
